### PR TITLE
Synchronize ticker and histogram metrics for Java API

### PR DESCRIPTION
**Pull request by adam-singer**
Updating the `HistogramType.java` and `TickerType.java` to expose and correct metrics for statistics callbacks.

Moved `NO_ITERATOR_CREATED` to the proper stat name and deprecated `NO_ITERATORS`.

### CODE REVIEW

**Comment by adam-singer** [comment]
doh

```
./java/./rocksjni/portal.h: In static member function ‘static rocksdb::Tickers rocksdb::TickerTypeJni::toCppTickers(jbyte)’:
./java/./rocksjni/portal.h:3740:12: error: case label value exceeds maximum value for type [-Werror]
       case 0x80:
```

**Comment by benclay** [comment]
Looks like we'll have to convert `jbyte` -> `jint`... Gross.

**Comment by adam-singer** [comment]
> Gross

I widened to `jchar`, could go `jint`. Are there any preferences here? Should we also widen `HistogramType` if not directly needed?

File: java/src/main/java/org/rocksdb/TickerType.java [hunk 1]
```diff
@@ -453,43 +454,267 @@
...
+     */
+    TXN_SNAPSHOT_MUTEX_OVERHEAD((char) 0x8A),
 
+    TICKER_ENUM_MAX((char) 0x8B);
 
-    private final byte value;
+    private final char value;
 
-    TickerType(final byte value) {
+    TickerType(final char value) {
```
Comment (benclay) [inline java/src/main/java/org/rocksdb/TickerType.java]: Hmm this is going to be a breaking change... @sagar0 what's our policy here? For the constructor we could have a `byte` variant that coerces to a char, but the getValue() accessor won't support that.

**Comment by benclay** [comment]
> I widened to `jchar`, could go `jint`.

I'd vote to go to jint unless jchar fits a uint32 which appears to be the type in statistics.h - no sense in doing this twice.

> Are there any preferences here? Should we also widen `HistogramType` if not directly needed?

Will let others vote here. @sagar0 ?

**Comment by adam-singer** [comment]
Seems the build was flaky on `NewJemallocNodumpAllocator`, re ran travis build on [adam-singer/rocksdb/builds/462039621|https://travis-ci.org/adam-singer/rocksdb/builds/462039621?utm_source=github_status&utm_medium=notification] without failures.

**Comment by adam-singer** [comment]
@sagar0 ping, any thoughts widening the type in `statistics.h` or getting this reviewed?

File: java/src/main/java/org/rocksdb/TickerType.java (hunk 1 above)
Comment (sagar0) [reply java/src/main/java/org/rocksdb/TickerType.java]: yeah, this is going to be a breaking change, and we don't allow breaking the API in minor versions. 
One BlobDB is not supported in RocksJava yet, we can may be get away by not adding BlobDB tickers and still keep within a byte for now?

**Comment by sagar0** [comment]
I think we should go with using a `jint` for `TickerType` and `HistogramType`, to conform to what the C++ side uses in statistics.h. But this will break the Java API now, which we can't allow until the next major version. 

As @benclay suggested one way is to introduce overloaded constructors for `TickerType` and `HistogramType`, and may be add new `getValueInt()` methods ... to not break existing users.

**Comment by adam-singer** [comment]
sgtm, I'll refactor to use `jint` and keep the `byte` contract in `TickerType` and `HistogramType` with an overloaded constructor for `int` exposed via `getValueInt`.

**Comment by sagar0** [comment]
@adamretter any concerns with this approach?

**Comment by adamretter** [comment]
Okay so @sagar0 do we *really* have a problem with backwards compatibility if we make changes here?

My thinking is that the internal value whether it is a `jbyte` or `jint` or whatever is only shared internally between the Java and C++ code across the JNI call. A user, is very unlikely to work with the raw value, instead they are likely to use the Enum type label.

At the moment `TickerType#getValue()` is *public* scope, it should likely have been *package-private* scope to stop users using the raw value.

Regardless. I can see another option that we have not considered. `jbyte` is a signed 8 bit type, so we have space for 255 values. In the C++ code the error seems to have been adding values about `0x80` as it is out of range because the `jbyte` is a signed type.

Why not just make use of the negative integer values that we can store in a signed type? The values themselves are unimportant! This gives us backwards compatibility with no changes needed.

**Comment by adam-singer** [comment]
@adamretter if we did that, what value would `TICKER_ENUM_MAX` have 0x7F? I think with the overloaded constructor approach that would also be similar issue. In terms of a solution, I sounds like it would work, feels a bit wonky semantics (for the interim) if anyone adds new stats with mappings. It would free up about `116`. If widening to `int`, that would give us more space then we could possibly need, also it maps naturally to the internal representation.

Since most of this concern is on breaking API, what is the cadence or goals for the next major releases? Would it be a palatable approach to do the widening of `jbyte` to `jint` near that time?

**Comment by adam-singer** [comment]
> > @adamretter if we did that, what value would `TICKER_ENUM_MAX` have 0x7F?
> 
> `TICKER_ENUM_MAX` is currently `0x5F`, so it should remain as that. The value itself is unimportant, however it is important to maintain the current value for backwards compatibility.

Seems in https://github.com/facebook/rocksdb/pull/4498/files it got jumbled around, will update that.

> 
> > I think with the overloaded constructor approach that would also be similar issue. In terms of a solution, I sounds like it would work, feels a bit wonky semantics (for the interim) if anyone adds new stats with mappings. It would free up about `116`. If widening to `int`, that would give us more space then we could possibly need, also it maps naturally to the internal representation.
> 
> So, if we want to maintain the current behaviour because as @sagar0 mentioned, i.e. we don't make API breaking changes during a minor or hotfix release, then I think we should stick with `jbyte`. We should also add a note to the javadoc for `TickerType#getValue()` that it will become package private in future.

Got it, will update review to reflect this.

> We should likely do that for all the Java Enums which are mappings to C++ enums, by sealing off their implementation from the API, we are then free to change the internal types however we like in future without issues such as this one.

Agree

> > Since most of this concern is on breaking API, what is the cadence or goals for the next major releases? Would it be a palatable approach to do the widening of `jbyte` to `jint` near that time?
> 
> I am not convinced we need to widen for the sake of doing so. In the next major release the enum getValue() methods should instead become package-private, allowing us to do any widening when required in future without breaking compatibility of the public API.

sgtm

**Comment by adam-singer** [comment]
@adamretter ptal

**Comment by sagar0** [comment]
@adamretter:
Thanks for looking into this. 
My original concern for the change was because of the return type of getValue(), as it was a public API. If we don't plan on exposing it to the user at all, with the current deprecation, that point is moot.

I am much less concerned about internal implementation details. While I personally don't like the use of negative values for the enums, its an internal detail and I don't mind the change (since it no longer spills into the public API).

**Comment by adamretter** [comment]
@sagar0 Can we get this merged now?

**Comment by adam-singer** [comment]
Checking in on any blockers to this

**Pull request closed without merging**